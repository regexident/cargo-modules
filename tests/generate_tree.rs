#[macro_use]
mod util;

mod smoke {
    test_cmd!(
        args: "generate tree",
        success: true,
        color_modes: ColorModes::ALL,
        project: smoke
    );
}

mod default {
    mod pass {
        test_cmds!(
            args: "generate tree",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate tree",
            success: false,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod lib {
    mod pass {
        test_cmds!(
            args: "generate tree \
                    --lib",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate tree \
                    --lib", // does not exist
            success: false,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }
}

mod bin {
    mod pass {
        test_cmds!(
            args: "generate tree \
                    --bin package_bin_target",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "generate tree \
                    --bin package_multi_target",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate tree \
                    --bin foobar", // does not exist
            success: false,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package {
    mod pass {
        test_cmds!(
            args: "generate tree \
                    --package package_lib_target",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "generate tree \
                    --package package_bin_target",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate tree \
                    --package foobar",
            success: false,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package_lib {
    mod pass {
        test_cmds!(
            args: "generate tree \
                    --package package_lib_target \
                    --lib",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_lib_target,
                workspace_single_package_lib_target,
                virtual_workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "generate tree \
                    --package package_multi_target \
                    --lib",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate tree \
                    --package package_bin_target \
                    --lib", // does not exist
            success: false,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );
    }
}

mod package_bin {
    mod pass {
        test_cmds!(
            args: "generate tree \
                    --package package_bin_target \
                    --bin package_bin_target",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "generate tree \
                    --package package_multi_target \
                    --bin package_multi_target",
            success: true,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate tree \
                    --package workspace-package \
                    --bin foobar", // does not exist
            success: false,
            color_modes: ColorModes::PLAIN,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod with_orphans {
    test_cmd!(
        args: "generate tree \
                --with-orphans",
        success: true,
        color_modes: ColorModes::PLAIN,
        project: smoke
    );
}

mod with_tests {
    test_cmd!(
        args: "generate tree \
                --with-tests",
        success: true,
        color_modes: ColorModes::PLAIN,
        project: smoke
    );
}

mod with_types {
    test_cmd!(
        args: "generate tree \
                --with-types",
        success: true,
        color_modes: ColorModes::PLAIN,
        project: smoke
    );
}
