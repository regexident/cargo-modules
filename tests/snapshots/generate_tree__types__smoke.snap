---
source: tests/generate_tree.rs
expression: output
---
STDERR:

STDOUT:

crate smoke
├── mod derives: pub(crate)
│   └── struct Dummy: pub(self)
├── mod functions: pub(crate)
│   ├── type Core: pub(self)
│   ├── type Crate: pub(self)
│   ├── struct Local: pub(self)
│   └── type Std: pub(self)
├── mod hierarchy: pub(crate)
│   └── mod lorem: pub(self)
│       ├── struct Lorem: pub(self)
│       ├── mod consectetur: pub(self)
│       │   ├── struct Consectetur: pub(self)
│       │   └── mod adipiscing: pub(self)
│       │       ├── struct Adipiscing: pub(self)
│       │       └── mod elit: pub(self)
│       │           └── struct Elit: pub(self)
│       ├── mod dolor: pub(self)
│       │   ├── struct Dolor: pub(self)
│       │   └── mod sit: pub(self)
│       │       ├── struct Sit: pub(self)
│       │       └── mod amet: pub(self)
│       │           └── struct Amet: pub(self)
│       └── mod ipsum: pub(self)
│           └── struct Ipsum: pub(self)
├── mod methods: pub(crate)
│   ├── type Core: pub(self)
│   ├── type Crate: pub(self)
│   ├── struct Dummy: pub(self)
│   ├── struct Local: pub(self)
│   └── type Std: pub(self)
├── mod orphans: pub(crate)
├── mod target: pub(crate)
│   └── struct Target: pub
├── mod uses: pub(crate)
│   └── mod cycle: pub(self)
│       ├── mod node_0: pub(self)
│       └── mod node_1: pub(self)
│           └── mod node_2: pub(self)
└── mod visibility: pub(crate)
    └── mod dummy: pub(self)
        ├── mod enums: pub(self)
        │   ├── enum PubCrate: pub(crate)
        │   ├── enum PubModule: pub(in crate::visibility)
        │   ├── enum PubPrivate: pub(self)
        │   ├── enum PubPublic: pub
        │   └── enum PubSuper: pub(super)
        ├── mod fns: pub(self)
        ├── mod kinds: pub(self)
        │   ├── enum Enum: pub(self)
        │   ├── mod Module: pub(self)
        │   ├── struct Struct: pub(self)
        │   ├── type TraitAlias: pub(self)
        │   ├── type TypeAlias: pub(self)
        │   └── union Union: pub(self)
        ├── mod mods: pub(self)
        │   ├── mod pub_crate: pub(crate)
        │   ├── mod pub_module: pub(in crate::visibility)
        │   ├── mod pub_private: pub(self)
        │   ├── mod pub_public: pub
        │   └── mod pub_super: pub(super)
        ├── mod structs: pub(self)
        │   ├── struct PubCrate: pub(crate)
        │   ├── struct PubModule: pub(in crate::visibility)
        │   ├── struct PubPrivate: pub(self)
        │   ├── struct PubPublic: pub
        │   └── struct PubSuper: pub(super)
        ├── mod traits: pub(self)
        │   ├── mod r#unsafe: pub(self)
        │   └── mod safe: pub(self)
        └── mod unions: pub(self)
            ├── union PubCrate: pub(crate)
            ├── union PubModule: pub(in crate::visibility)
            ├── union PubPrivate: pub(self)
            ├── union PubPublic: pub
            └── union PubSuper: pub(super)

